community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/SemanticState.scala

```diff
@@ -186,7 +187,7 @@ import org.neo4j.cypher.internal.compiler.v2_3.SemanticState.ScopeLocation
 case class SemanticState(currentScope: ScopeLocation,
                          typeTable: ASTAnnotationMap[ast.Expression, ExpressionTypeInfo],
                          recordedScopes: ASTAnnotationMap[ast.ASTNode, Scope],
-                         notificationLogger: InternalNotificationLogger = devNullLogger) {
+                         notifications: Set[InternalNotification] = Set.empty) {
```

Comment:
`SemanticState` is a monad and should not contain mutable state (the InternalNotificationLogger is mutable). @boggle and I made this discovery and changed it while we were passing through.

The `devNullLogger` still has its use on line 31 in `SemanticChecker` above, where notifications are passed to the logger. If that's the `devNullLogger`, notifications will be discarded.